tests/unit/s2n_client_alpn_extension_test.c

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+
+#include "tls/extensions/s2n_client_alpn.h"
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    const char *protocols[] = { "protocol1", "protocol2", "protocol3" };
+    const uint8_t protocols_count = s2n_array_len(protocols);
+
+    struct s2n_connection *conn;
+    EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+
+    /* Test should_send */
+    {
+        EXPECT_SUCCESS(s2n_connection_set_protocol_preferences(conn, NULL, 0));
+        EXPECT_FALSE(s2n_client_alpn_extension.should_send(conn));
+
+        EXPECT_SUCCESS(s2n_connection_set_protocol_preferences(conn, protocols, protocols_count));
+        EXPECT_TRUE(s2n_client_alpn_extension.should_send(conn));
+    }
+
+    /* Test send */
+    {
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+        EXPECT_SUCCESS(s2n_connection_set_protocol_preferences(conn, protocols, protocols_count));
+
+        EXPECT_SUCCESS(s2n_client_alpn_extension.send(conn, &stuffer));
```

Comment:
Can we add some negative test cases for send and recv?

Reply:
There isn't really a negative case for send, but I'll add a few more for recv.